c++/src/Reader.cc

```diff
@@ -897,6 +897,27 @@ namespace orc {
     return REDUNDANT_MOVE(postscript);
   }
 
+  // ORC-317: check that indices in the type tree are valid, so we won't crash
+  // when we convert the proto::Types to TypeImpls.
+  void checkProtoTypeIds(int &index, const proto::Footer &footer) {
```

Comment:
I'd suggest only passing in the footer here and check all of the types in a loop.